### PR TITLE
<format>: Assume UTF-8 format strings when execution charset is UTF-8

### LINKED ISSUE #1820: <format>: Incorrect handling of UTF-8 encoded format strings
**Describe the bug**

`std::format` throws on formatting a valid UTF-8 string when the literal (execution) encoding is UTF-8 and the locale encoding is Shift-JIS (and possibly other cases). This is wrong because format strings are almost always literals and therefore the locale encoding shouldn't affect parsing.

**Command-line test case**

```
C:\Temp>type test.cc
#include <format>
#include <iostream>

int main() {
  setlocale(LC_ALL, ".932");
  try {
    auto s = std::format("\xe2\x82\xa0{}", 42);
    std::cout << (s == "\xe2\x82\xa0" "42") << "\n";
  } catch (const std::exception& e) {
    std::cout << e.what() << "\n";
  }
}

C:\Temp>cl /I STL/stl/inc /std:c++latest /utf-8 test.cc
Microsoft (R) C/C++ Optimizing Compiler Version 19.29.29917 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

/std:c++latest is provided as a preview of language features from the latest C++
working draft, and we're eager to hear about bugs and suggestions for improvements.
However, note that these features are provided as-is without support, and subject
to changes or removal as the working draft evolves. See
https://go.microsoft.com/fwlink/?linkid=2045807 for details.

test.cc
test.cc(6): warning C4530: C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc
Microsoft (R) Incremental Linker Version 14.29.29917.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:test.exe
test.obj

C:\Temp>test
Invalid encoded character in format string.
```

**Expected behavior**
The test program should print 1.

**STL version**
    ```
    https://github.com/microsoft/STL/commit/4d7d4f1
    ```

### PR DESCRIPTION
**Pull request by statementreply**
When execution charset is UTF-8, assume that format strings are encoded in UTF-8, not in the active code page.

This PR only attempts to detect UTF-8.
- UTF-8 is _the_ modern and most important text encoding.
- A Windows program has always been doomed when it uses non-Unicode encodings, and the developers and the users assume different code pages.

Fixes #1820.

### CODE REVIEW

File: tests/std/tests/P0645R10_text_formatting_utf8/test.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        assert(format("{:\xf0\x9f\x8f\x88^4.2}", "1") == "\xf0\x9f\x8f\x88\x31\xf0\x9f\x8f\x88\xf0\x9f\x8f\x88"sv);
+        assert(format("{:\xf0\x9f\x8f\x88>4.2}", "1") == "\xf0\x9f\x8f\x88\xf0\x9f\x8f\x88\xf0\x9f\x8f\x88\x31"sv);
+    }
+
+    {
+        try {
+            (void) format("{:\x9f\x8f\x88<10}"sv, 42); // Bad fill character encoding: missing lead byte before \x9f
+            assert(false);
+        } catch (const format_error&) {
+        }
```
Comment (miscco) [inline tests/std/tests/P0645R10_text_formatting_utf8/test.cpp:28]: We could check the error message here

**Review by barcharcraz — CHANGES_REQUESTED** [review]

File: stl/inc/format [hunk 2]
```diff
@@ -417,55 +417,79 @@ _NODISCARD constexpr const _CharT* _Parse_arg_id(
     throw format_error("Invalid format string.");
 }
 
+inline constexpr bool _Execution_charset_is_utf8 = [] {
```
[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/format:420]: This is not fully correct. This is an inline variable. There `inline` guarantees that there is only one instance of it program wide.

It is a bit confusing here, as it is an immediately invoked lambda expression. So it is more akin of  `inline constexpr bool traits_v = ....`

[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/format:420]: Nope, because variable

File: stl/inc/format [hunk 3]
```diff
@@ -417,55 +417,79 @@ _NODISCARD constexpr const _CharT* _Parse_arg_id(
...
+#pragma warning(disable : 4566) // character represented by universal-character-name '\u4E00' cannot be represented in
+                                // the current code page
+#pragma warning(disable : 6201) // Index '2' is out of valid index range '0' to '1' for possibly stack allocated buffer
+                                // '_Test_char'
+#pragma warning(disable : 6239) // (<non-zero constant> && <expression>) always evaluates to the result of <expression>.
+                                // Did you intend to use the bitwise-and operator?
+    constexpr char _Test_char[] = "\u4e00";
+    return sizeof(_Test_char) == 4 && _Test_char[0] == '\xe4' && _Test_char[1] == '\xb8' && _Test_char[2] == '\x80';
+#pragma warning(pop)
+}();
```
[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/format:432]: We could make the lambda a constexpr funcntion, but we will still need the variable

File: stl/inc/format [hunk 4]
```diff
@@ -417,55 +417,79 @@ _NODISCARD constexpr const _CharT* _Parse_arg_id(
...
+
+_NODISCARD inline int _Utf8_code_units_in_next_character(const char* const _First, const char* const _Last) noexcept {
```
Comment (miscco) [inline stl/inc/format:434]: from what I see nothing in this function prevents it from being `constexpr`

However, I believe it only makes sense at runtime. So should we add a comment that this is intentionally *not* constexpr

File: stl/inc/format [hunk 5]
```diff
@@ -417,55 +417,79 @@ _NODISCARD constexpr const _CharT* _Parse_arg_id(
...
+                mbstate_t _St{};
+                const auto _Len   = static_cast<size_t>(_Last - _First);
+                const int _Result = _Mbrtowc(&_Wide, _First, _Len, &_St, &_Cvt);
+                if (_Result > 0) {
+                    return _Result;
+                } else if (_Result < 0) { // invalid or incomplete encoded character
+                    return -1;
+                } else { // next code unit is '\0'
+                    return 1;
+                }
```
Comment (miscco) [inline stl/inc/format:486]: I am wondering whether we should encapsulate this into a function to give it a proper name. That way we could hide the complexity of the individual subfunction and keep `_Code_units_in_next_character` easy to reason about

**Review by miscco — APPROVED** [review]
From my inexperienced point of view this looks good. I would like to cleanup `_Code_units_in_next_character` so that it only defers to subfunctions but this is purely a style thing which can / should be disregarded

File: stl/inc/format (hunk 4 above)
Comment (statementreply) [reply stl/inc/format]: Made it `constexpr`

File: tests/std/tests/P0645R10_text_formatting_utf8/test.cpp (hunk 1 above)
Comment (barcharcraz) [reply tests/std/tests/P0645R10_text_formatting_utf8/test.cpp:28]: eagh, the value of checking the error messages is very low.

**Review by barcharcraz — APPROVED** [review]

File: tests/std/tests/P0645R10_text_formatting_utf8/test.cpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <format>
+#include <locale>
```
Comment (StephanTLavavej) [inline tests/std/tests/P0645R10_text_formatting_utf8/test.cpp:6]: Ditto, should instead be `<clocale>`.

File: tests/std/tests/P0645R10_text_formatting_legacy_text_encoding/test.cpp [hunk 7]
```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <format>
+#include <locale>
```
Comment (StephanTLavavej) [inline tests/std/tests/P0645R10_text_formatting_legacy_text_encoding/test.cpp:6]: Should instead include `<clocale>` for `setlocale()`.

**Review by StephanTLavavej — COMMENTED** [review]
Looks good - I'll push changes for my minor comments here, and a couple more to mitigate merge conflicts with the commits I recently pushed to `merging_format`.

**Review by StephanTLavavej — APPROVED** [review]

**Pull request merged**